core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,555 @@
+      if (substream != winningSubstream) {
+        substream.cancel(Status.CANCELLED.withDescription(CANCELLED_BECAUSE_HEDGE_COMMITTED));
+      }
+    }
+
+    return true;
+  }
+
+  private void retry() {
+    ClientStream substream = newSubstream();
+
+    // TODO(zdapeng): update "grpc-retry-attempts" header
+
+    drain(substream);
+  }
+
+  private ClientStream newSubstream() {
+    ClientTransport transport =
+        clientTransportProvider.get(new PickSubchannelArgsImpl(method, headers, callOptions));
+    Context origContext = context.attach();
+    try {
+      return transport.newStream(method, headers, callOptions);
+    } finally {
+      context.detach(origContext);
+    }
+  }
+
+  private void drain(ClientStream substream) {
+    int index = 0;
+    int chunk = 0x80;
+    List<BufferEntry> list = null;
+    Status status;
+
+    while (true) {
+      synchronized (lock) {
+        int stop = Math.min(index + chunk, state.buffer.size());
+        if (list == null) {
+          list = new ArrayList<BufferEntry>(stop - index);
+        }
+        list.clear();
+        list.addAll(state.buffer.subList(index, stop));
+        index = stop;
+      }
+
+      for (BufferEntry bufferEntry : list) {
+        State savedState = state;
+        if (savedState.cancellingStatus != null) {
+          substream.cancel(savedState.cancellingStatus);
+          return;
+        }
+        if (savedState.winningSubstream != null && savedState.winningSubstream != substream) {
+          // committed but not me
+          substream.cancel(Status.CANCELLED.withDescription(CANCELLED_BECAUSE_HEDGE_COMMITTED));
+          return;
+        }
+        bufferEntry.runWith(substream);
+      }
+
+      synchronized (lock) {
+        if (index != state.buffer.size()) { // I'm not drained
+          continue;
+        }
+
+        if (state.cancellingStatus != null) {
+          status = state.cancellingStatus;
+          break;
+        }
+        if (state.winningSubstream != null && state.winningSubstream != substream) {
+          // committed but not me
+          status = Status.CANCELLED.withDescription(CANCELLED_BECAUSE_HEDGE_COMMITTED);
+          break;
+        }
+
+        state = state.drained(substream);
+        return;
+      }
+    }
+
+    substream.cancel(status);
+  }
+
+  /** Starts the first PRC attempt. */
+  @Override
+  public void start(ClientStreamListener listener) {
+    masterListener = listener;
+    delayedClientTransport.addUncommittedRetriableStream(this);
+
+    class StartEntry implements BufferEntry {
+      @Override
+      public void runWith(ClientStream substream) {
+        substream.start(new Sublistener(substream));
+      }
+    }
+
+    synchronized (lock) {
+      state.buffer.add(new StartEntry());
+    }
+
+    ClientStream substream = newSubstream();
+    drain(substream);
+
+    // TODO(zdapeng): schedule hedging if needed
+  }
+
+  @Override
+  public void cancel(Status reason) {
```

Comment:
It seems like this should close masterListener, at least in some cases (like if we are between retry-attempts).